src/main/java/io/supertokens/totp/Totp.java

```diff
@@ -0,0 +1,301 @@
+        }
+
+        // Check if the code has been previously used by the user and it was valid (and
+        // is still valid). If so, this could be a replay attack. So reject it.
+        if (isValid) {
+            for (TOTPUsedCode usedCode : usedCodes) {
+                // One edge case is that if the user has 2 devices, and they are used back to
+                // back (within 90 seconds) such that the code of the first device was
+                // regenerated by the second device, then it won't allow the second device's
+                // code to be used until it is expired.
+                // But this would be rare so we can ignore it for now.
+                if (usedCode.code.equals(code) && usedCode.isValid
+                        && usedCode.expiryTime > System.currentTimeMillis()) {
+                    isValid = false;
+                    matchingDevice = null;
+                }
+            }
+        }
+
+        // Insert the code into the list of used codes:
+        long now = System.currentTimeMillis();
+        int invalidCodeExpirySec = Config.getConfig(main).getTotpInvalidCodeExpiryTime(); // (Default 30 mins)
+        int expireInSec = isValid ? matchingDevice.period * (2 * matchingDevice.skew + 1) : invalidCodeExpirySec;
+
+        TOTPUsedCode newCode = new TOTPUsedCode(userId, code, isValid, now + 1000 * expireInSec, now);
+        totpStorage.insertUsedCode(newCode);
+
+        if (!isValid) {
+            throw new InvalidTotpException();
+        }
+    }
+
+    public static boolean verifyDevice(Main main, String userId, String deviceName, String code)
+            throws StorageQueryException, TotpNotEnabledException, UnknownDeviceException, InvalidTotpException,
+            LimitReachedException {
+        // Here boolean return value tells whether the device has been
+        // newly verified (true) OR it was already verified (false)
+
+        TOTPSQLStorage totpStorage = StorageLayer.getTOTPStorage(main);
+        TOTPDevice matchingDevice = null;
+
+        // Check if the user has any devices:
+        TOTPDevice[] devices = totpStorage.getDevices(userId);
+        if (devices.length == 0) {
+            throw new TotpNotEnabledException();
+        }
+
+        // Check if the requested device exists:
+        for (TOTPDevice device : devices) {
+            if (device.deviceName.equals(deviceName)) {
+                matchingDevice = device;
+                if (device.verified) {
+                    return false; // Was already verified
+                }
+                break;
+            }
+        }
+
+        // No device found:
+        if (matchingDevice == null) {
+            throw new UnknownDeviceException();
+        }
+
+        checkAndStoreCode(main, totpStorage, userId, new TOTPDevice[] { matchingDevice }, code);
+        // Will reach here only if the code is valid:
+        totpStorage.markDeviceAsVerified(userId, deviceName);
+        return true; // Newly verified
```

Comment:
There is a race condition here where two requests mark the device as verified at the same time. In this case, the APIs will both return true. So we should check if `markDeviceAsVerified` marked the device as verified or just did a no-op?

Reply:
And how to do that? Do another query or just use `UPDATE SET ... WHER .... AND verified = False`?